### PR TITLE
Use correct priority to compute allocated resources

### PR DESCRIPTION
**Pull request by zuqq**
The issue that I'm addressing here is that `totalAllocationByPoolAndQueue` uses the `Priority` field of `JobSchedulingInfo`:

https://github.com/armadaproject/armada/blob/f3fa267c875dd7908a7efc6fd62dd38a8a014a77/internal/scheduler/scheduling_algo.go#L544-L551

… even though that field represents the per-queue priority of the job, not its priority class:

https://github.com/armadaproject/armada/blob/f3fa267c875dd7908a7efc6fd62dd38a8a014a77/internal/scheduler/schedulerobjects/schedulerobjects.pb.go#L621-L622

It might be worth having a separate type for the priority number associated with each priority class, in order to distinguish it from the other integers that are floating around.

### CODE REVIEW

File: internal/scheduler/scheduling_algo.go [hunk 1]
```diff
@@ -57,13 +57,11 @@ func NewFairSchedulingAlgo(
...
-		for _, p := range config.Preemption.PriorityClasses {
-			priorities = append(priorities, p.Priority)
-		}
-	} else {
-		priorities = append(priorities, 0)
+	priorityClasses := config.Preemption.PriorityClasses
+	// The comment on `DefaultPriorityClass` suggests that we should never take
+	// this branch, but we are being extra cautious here.
+	if len(config.Preemption.PriorityClasses) <= 0 {
+		priorityClasses = map[string]configuration.PriorityClass{config.Preemption.DefaultPriorityClass: {Priority: 0}}
```
Comment (severinson) [inline internal/scheduler/scheduling_algo.go:64]: We should error on no PCs imo.

File: internal/scheduler/scheduling_algo.go [hunk 2]
```diff
@@ -543,10 +541,12 @@ func (l *FairSchedulingAlgo) totalAllocationByPoolAndQueue(executors []*schedule
 			}
 			jobSchedulingInfo := job.JobSchedulingInfo()
 			if jobSchedulingInfo != nil {
-				allocation.AddResourceList(
-					int32(jobSchedulingInfo.Priority),
-					jobSchedulingInfo.GetTotalResourceRequest(),
-				)
+				priorityClass, ok := l.priorityClasses[jobSchedulingInfo.PriorityClassName]
```
Comment (severinson) [inline internal/scheduler/scheduling_algo.go:544]: We shouldn't schedule this job. Either skip the job entirely or return an error.

Comment (zuqq) [reply internal/scheduler/scheduling_algo.go:544]: This function is called at the start of the scheduling round, when constructing a `fairShareSchedulingContext`:

https://github.com/armadaproject/armada/blob/f3fa267c875dd7908a7efc6fd62dd38a8a014a77/internal/scheduler/scheduling_algo.go#L289-L290

If I understand correctly, the jobs that we are looping over have already been scheduled; in particular, encountering an unknown priority class here means that the scheduler is in an inconsistent state.

This function could return an error instead, but I'm worried about the implications of making this change. If we abort the scheduling round here, wouldn't the next leader just run into the same issue?

Comment (zuqq) [reply internal/scheduler/scheduling_algo.go:544]: Actually, returning an error would be consistent with some of the other error handling in `scheduleOnExecutor`:

https://github.com/armadaproject/armada/blob/f3fa267c875dd7908a7efc6fd62dd38a8a014a77/internal/scheduler/scheduling_algo.go#L373-L377

On the other hand, `constructNodeDb` seems to prefer to just log and carry on:

https://github.com/armadaproject/armada/blob/f3fa267c875dd7908a7efc6fd62dd38a8a014a77/internal/scheduler/scheduling_algo.go#L435-L442

**Review by severinson — APPROVED** [review]

**Pull request merged**